src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -814,6 +818,80 @@ object Genotype {
     gt
   }
 
+  def readBiallelicDosageGenotype(isDosage: Boolean, a: ByteIterator): Double = {
+    val nAlleles = 2
+    val isBiallelic = true
+
+    val flags = a.readULEB128()
+
+    val gt: Int =
+      if (flagHasGT(isBiallelic, flags)) {
+        if (flagStoresGT(isBiallelic, flags))
+          flagGT(isBiallelic, flags)
+        else
+          a.readULEB128()
+      } else
+        -1
+
+    var count = 0
+
+    if (flagHasAD(flags)) {
+      if (flagSimpleAD(flags)) {
+        count += 1
+        val p = Genotype.gtPair(gt)
+        if (p.j != p.k)
+          count += 1
+      } else
+        count += nAlleles
+    }
+
+    if (flagHasDP(flags) && !(flagHasAD(flags) && flagSimpleDP(flags)))
+      count += 1
+
+    a.skipLEB128(count)
```

Comment:
`count` is not used after the skip instruction. the fact that this is easy to miss is part of why I'm strongly inclined to abstract these bits. Yes they'd need `isBiallelic` and `a` as well.

The iterator should probably just have a `GenotypeReader` object which has its fingers on `isBiallelic` (which, AFAIK, should have one value for the whole iterator) and the byte iterator. Usage might look like:

```scala
gr.startGenotype()
val gt = gr.readGt()
gr.skipAdDp()
val dosage = gr.readDosage()
gr.skipGQ()
```

It's not ideal:
 - it's easy to screw up and forget one of these steps (but easier to verify that all necessary steps are present than in the current state)
 - we repeat `gr` on each line
 - we can get the order wrong (but again, easier to verify that order is correct than in present state)

There are better solutions [1] that require more engineering that I suspect isn't worth the time right now.

[1] One idea: have some way to write `Reader(gt = true, adDp = false, dosage = true, gq = false)` (there's one `Reader` object for the entire `ByteIterator`)  which just has a series of `if`s that skip or read the field, populating fields (e.g. `gt`, `dosage`) on the `Reader`, and usage looks like:
```scala
gr.read()
val dosage = gr.dosage
```